src/playground-project.ts

```diff
@@ -465,74 +457,39 @@ export class PlaygroundProject extends LitElement {
-    await this._compileResultPromise;
-    const compiledUrl = new URL(name, window.origin).href;
-    const compiledContent = this._compiledFiles?.get(compiledUrl);
-    if (compiledContent !== undefined) {
+    if (this._build === undefined) {
       return {
-        name,
-        label: this._files?.find((f) => f.name === name)?.label,
-        content: compiledContent,
-        contentType: 'application/javascript',
+        status: /* Service Unavailable */ 503,
+        body: 'Playground build not started',
       };
-    } else {
-      const file = this._files?.find((f) => f.name === name);
-      if (file === undefined) {
-        return {status: 404, body: 'Playground file not found'};
-      }
-      return file;
     }
+    return this._build.getFile(name);
   }
 
+  /**
+   * Build this project immediately, cancelling any previous build.
+   */
   async save() {
-    // Clear in case a save is explicitly requested while a timer is already
-    // running.
-    const compileId = nextCompileId++;
-    this._compileId = compileId;
-    const compileStale = () => compileId !== this._compileId;
-
-    this._clearSaveTimeout();
-    this._compiledFiles = undefined;
+    this._build?.cancel();
+    const build = new PlaygroundBuild(() => {
+      this.dispatchEvent(new CustomEvent('diagnosticsChanged'));
+    });
+    this._build = build;
     this.dispatchEvent(new CustomEvent('compileStart'));
-    if (this._files !== undefined) {
-      const workerApi = await this._deferredTypeScriptWorkerApi.promise;
-      if (compileStale()) {
-        return;
-      }
-      this._compileResultPromise = workerApi.compileProject(
-        this._files,
-        this._importMap,
-        proxy((slowDiagnostics: Map<string, Array<Diagnostic>>) => {
-          if (compileStale() || slowDiagnostics.size === 0) {
-            return;
-          }
-          this._diagnostics =
-            this._diagnostics !== undefined
-              ? mergeArrayMaps(this._diagnostics, slowDiagnostics)
-              : slowDiagnostics;
-          this.dispatchEvent(new CustomEvent('diagnosticsChanged'));
-        })
-      );
-      const result = await this._compileResultPromise;
-      if (compileStale()) {
-        return;
-      }
-      this._compiledFiles = result?.files;
-      this._diagnostics = result?.diagnostics;
-    } else {
-      this._compileResultPromise = Promise.resolve(undefined);
+    const workerApi = await this._deferredTypeScriptWorkerApi.promise;
+    if (build.state() !== 'active') {
+      return;
     }
-    this.dispatchEvent(new CustomEvent('compileDone'));
-    this.dispatchEvent(new CustomEvent('diagnosticsChanged'));
-  }
-
-  private _saveTimeoutId?: ReturnType<typeof setTimeout> = undefined;
-
-  private _clearSaveTimeout() {
-    if (this._saveTimeoutId !== undefined) {
-      clearTimeout(this._saveTimeoutId);
-      this._saveTimeoutId = undefined;
+    workerApi.compileProject(
```

Comment:
Eventually it'd be nice to enable incremental builds for build-as-you-type and generally taking less time> Especially as deps won't change.

Reply:
Yep would like to tackle that sometime, after all this refactoring is done.